### PR TITLE
End coinjoin when anonymity reached

### LINKED ISSUE #7471: Coinjoin is still running even though the required level of anonymity of all utxos has been reached
During the feature testing, I had 5000 non-private sats on my Coinjoin account. 
After two rounds of CJ all my UTXOs has reached desired anonymity level (10) - I had no non-private sats left- however coinjoin still kept running and it kept pairing coins with rounds (and failing) in the endless loop.

Trezor Suite commit hash a7d89a929c5f22d6b80707269bd1d45900862dc1

Expected behavior is that coinjoin stops after desired level of anonymity is reached for all UTXOs of related account.

Screenshot 2023-01-26 at 16 00 30

Screenshot 2023-01-26 at 15 59 44

### PR DESCRIPTION
**Pull request by komret**
## Description

f2865d95cfd401863ec9467aa7655c601e5d1aae - update account at the end of each coinjoin round and check utxos; end coinjoin session and show success modal if there is nothing more to coinjoin (i.e all UTXOs with large enough amount have reached target anonymity).

- I noticed that the event for the final round phase is fired twice. I don't think it causes any problems so far, but can it be prevented somehow?

7ea390106e0032a75cafc91e3757399d9114e177 - end coinjoin session when all non-private UTXOs are spent during a (paused) coinjoin session. I added this to an effect in `CoinjoinStatus` for simplicity as the check happens there already and the component opens right after spending coins.

## Related Issue

Resolve #7471

## Screenshots:
Screenshot 2023-02-21 at 10 49 35

### CODE REVIEW

**Comment by komret** [comment]
@szymon, could you please check the changes I made [here](https://github.com/trezor/trezor-suite/pull/7680/commits/f2865d95cfd401863ec9467aa7655c601e5d1aae#diff-d433028b82fbafb54eedda1a2aae2b53b8f53e1fcd4cf71b29e21b060fdb0eb9)? I am not 100% sure updating the account here is correct and I noticed the event for final phase firing twice.

File: packages/suite/src/actions/wallet/coinjoinAccountActions.ts [hunk 1]
```diff
@@ -571,19 +564,28 @@ export const pauseCoinjoinSessionByDeviceId =
...
+                            {
+                                symbol: account.symbol,
+                            },
+                        ),
                     );
                 }
-                // get @trezor/coinjoin client if available
-                const client = getCoinjoinClient(account.symbol);
+                const hasRunningSession = selectIsAccountWithSessionByAccountKey(
+                    getState(),
```
Comment (szymonlesisz) [inline packages/suite/src/actions/wallet/coinjoinAccountActions.ts]:
```suggestion
                    state,
```

File: packages/suite/src/actions/wallet/coinjoinClientActions.ts [hunk 2]
```diff
@@ -257,38 +267,48 @@ export const onCoinjoinRoundChanged =
...
+                            await dispatch(coinjoinAccountActions.fetchAndUpdateAccount(account));
+                        }
+                    }),
+                );
 
-                        return !isZero(notAnonymized);
-                    });
+                const [accountsReachingAnonymity, accountsNotReachingAnonymity] = arrayPartition(
+                    coinjoinAccountsWithSession,
+                    ({ key }) => selectAreUtxosTooSmallByAccountKey(getState(), key),
```
Comment (matejkriz) [inline packages/suite/src/actions/wallet/coinjoinClientActions.ts]: please rename the selector

Comment (komret) [reply packages/suite/src/actions/wallet/coinjoinClientActions.ts]: 65a560cac224bb9a6375571ac3b08c02b6310d62

**Review by szymonlesisz — DISMISSED** [review]
i think it should be done in `updateClientAccount` action. anonymity of utxos is not calculated after phase change but after account history change

**Comment by szymonlesisz** [comment]
> I noticed that the event for the final round phase is fired twice. I don't think it causes any problems so far, but can it be prevented somehow?

yes, it's intended `CoinjoinClient` emits round change event twice: before and after processing round, this is why all suite actions are wrapped in [condition with comment](https://github.com/trezor/trezor-suite/blob/develop/packages/suite/src/actions/wallet/coinjoinClientActions.ts#L242-L244)

**Comment by komret** [comment]
I will move the check to `updateClientAccount`

> i think it should be done in `updateClientAccount` action. anonymity of utxos is not calculated after phase change but after account history change

OK, will do.

**Comment by komret** [comment]
I moved the check for anonymity to `updateClientAccount` action as proposed by @szymonlesisz. There is a problem, though: When the account reaches target anonymity during the last possible round (`signedRounds.length === maxRounds`), "More rounds needed" is shown instead of the success modal.

Another way this could be approached is to move the check for `maxRounds` to `updateClientAccount` as well, but that would mean starting a new round above the limit which is weird.

A proper solution seems to be introducing a new session phase of coinjoin evaluation between the end of the last permitted round and account update - I am leaving that for discussion and follow-up issue.

Do you have any more suggestions, @szymonlesisz?

832802f09b55dae49cb8682d6ef45938461e95b2 removes the logic for ending the session when UTXOs are spent as this case should now also be handled by `updateClientAccount`.

File: packages/suite/src/actions/wallet/coinjoinAccountActions.ts (hunk 1 above)
Comment (komret) [reply packages/suite/src/actions/wallet/coinjoinAccountActions.ts]: d7d059cb2376efc56c2a83a7cfc1963bdfe8a816

**Review by matejkriz — APPROVED** [review]
🎉 
image

**Pull request merged**